### ISSUE TITLE
Calculation of Counter T1 of P1 Smart meter is not correct

### ISSUE THREAD

**Issue by FireWizard52** [domoticz/domoticz#3896]
It looks that the calculation of the Counter T1 (Usage) is not correct.
All other counters (T2, R1 and R2) calculate exactly the correct values.
The value of the day should be the sum of the value of the previous day and the usage of the day.

Example:
January 1: Counter T1 (7551,166)  + Usage T1 20,448 should result in:
January 2: Counter T1 (7571,614) instead of 7571,796.

For more information: See https://www.domoticz.com/forum/viewtopic.php?f=6&t=31192
Posts of erem and FireWizard, February 6, 2020

Running: Domoticz 4.11670
Rasbian Buster on RPi2B

**Comment by gizmocuz** [comment]
That is odd as there is no 'calculation' being made , we use the direct counter values

It is stored in the database every (5) minutes here:

void CSQLHelper::UpdateMultiMeter()

and at the end of the day:

void CSQLHelper::AddCalendarUpdateMultiMeter()

you should be able to verify this in the database tables

**Comment by FireWizard52** [comment]
Thanks for the reply.
I will check tomorow and report.

**Comment by gordonb3** [comment]
This is a long known shortcoming, I wouldn't call it a defect perse. A day in Domoticz is not 24 hours because every 24 hour period you can think of contains at least one time value that has a different date, causing it to fall outside the query range used in these calculations. Because the values are stored in 5 minute intervals a day therefore becomes a period of 23 hours and 55 minutes. Typically this will result in a deviation of around 10Wh but of course it might also be close to a full kWh if you're charging an EV.

**Comment by FireWizard52** [comment]
After the forum reaction of gordonb3 and the fact that it is obviously a long known issue, it doesn't make sense for me to look into the database, so I wont do that.
As it is clear to me that it will only happen with Usage T1, it is the question, do we want to solve it?

**Comment by gordonb3** [comment]
Need to note that I have not looked at this particular implementation for quite some time. As such I'm not completely sure whether only data from the shortlog is used to calculate these daily values or if it uses the device state value as the last known value. Either way there appears to be a time difference between the time that this calculation is based on and the creation of the first shortlog entry for the next day. Whether that's 5 minutes, 10 seconds (DSMR standard meter) or 1 second (ESMR 5.x meter) doesn't really matter as the difference will always be notable.

**Comment by FireWizard52** [comment]
Please see the recent posts:
https://www.domoticz.com/forum/viewtopic.php?f=6&t=31192

**Comment by rwaaren** [comment]
> Please see the recent posts:
> https://www.domoticz.com/forum/viewtopic.php?f=6&t=31192

can this issue considered solved now ?

**Comment by FireWizard52** [comment]
After update to V4.11696 the counter Counter T1 seems to be correct.
However only tested for one day.

Thanks for all support

**Comment by FireWizard52** [comment]
The update from Sunday to Monday was correct but Monday to Tuesday and Tuesday to Wednesday are still not correct.
See screenshot https://www.domoticz.com/forum/viewtopic.php?f=6&t=31192

**Comment by gordonb3** [comment]
I was afraid that might happen, but as it didn't occur on my machine (which is still correct) I figured that the order of operations prevented the possible race condition.

Question: are you running Domoticz on a multi-threaded machine? If so, try binding the service to a single CPU thread (`taskset` command on linux). This is for verification only, I'm not suggesting this as a permanent fix.

**Comment by FireWizard52** [comment]
See my forum reply as well.
Indeed I'm running Domoticz on a quad-core CPU (Raspberry Pi 2B+)

I have set Domoticz to use core 3 only from now on.
Lets wait 3 more days, if this gives a possible cause.

**Comment by FireWizard52** [comment]
See the forum as well for a screenshot.
https://www.domoticz.com/forum/viewtopic.php?f=6&t=31192&start=20

The test results so far are positive and it looks that this might be a possible cause.

**Comment by gordonb3** [comment]
This confirms my suspicion. So here's what's happening. Domoticz has a number of fixed events, next to the ones that you may set as a user. One of these events is an addition to the shortlog, which copies the current registered status of a device every 5 minutes. The other member in this conflict is an update to the daily logs, "calendar_(multi)meter", which happens at midnight and that is also a 5 minute event. The race condition occurs because the latter event takes it information from the shortlog and thus requires the 5 minute event to have been completed to hold accurate information. Due to the order of firing this runs correct if your process runs single threaded, but on a multicore CPU they will run in parallel and the midnight event data query may precede the 5 minute event data update.

Effectively this means that the midnight event should be delayed. @gizmocuz input required.

**Comment by gizmocuz** [comment]
And this all to get a 5 minute correction ?

**Comment by gordonb3** [comment]
More to allow the math to work out, so that day start + usage becomes equal to next day start.

**Comment by gizmocuz** [comment]
So how many watts (kW) is this going to differ ?

**Comment by gizmocuz** [comment]
Do mind that we use the actual meter values, so things (few watts) not calculated yesterday will be corrected the next day

**Comment by gordonb3** [comment]
I'd say up to a full kWh is feasible.

And no. The daily meter reports in `meter_calendar` and `multimeter_calendar` are in fact composed from the shortlog only and these records are never updated once added, so there is no correction afterwards.

**Comment by FireWizard52** [comment]
@Gizmocuz wrote: 

> So how many watts (kW) is this going to differ ?

This depends on how you calculate per day/month/year.
In my case, it varies per day, but it differs 3-4 kWh per month and so approx 40 kWh per year.
All other values, like T2, R1 and R2 are perfectly correct and so is the gas consumption.

We know now, how the difference occurs, but in the future questions will return.
We also know, what the cause is and I suggest, if it is not too complicated, to correct it.

**Comment by gizmocuz** [comment]
@FireWizard52 , that is strange, as in the report (for the month) the total month usage could/should be calculated by the counter values and it should be correct, it can not be a few Kwh per month/year
For example month usage =  first value of next month - first value of month
There is no need to summarize all day values to produce a month total

**Comment by gordonb3** [comment]
That statement does not appear to be accurate.

I find the Angular code quite difficult to follow, however what I do see is that the year report is also based on day to day data rather than data that was already consolidated in webserver. Since the only place where I see the cost values being applied is on day records the total costs per month must be the sum of all these calculated values that were added to the day records, which are all missing the last 5 minutes. I think this occurs in the `getGroupStats` function of `EnergyMultiCounterReport.js` and it looks like this function also accumulates the registered usage and return values rather than from the counter values.

**Comment by gizmocuz** [comment]
Please have a look in webserver.cpp where the graph data is generated

**Comment by gordonb3** [comment]
I have. It is actually pretty weird because the code subtracts the usage from the counter value to get to the day start counter value. I suppose it is due to some backward compatibility that the database contains `max(counter)` while the report shows `min(counter)`.

But it is actually much simpler to use the web developer tools in Firefox. If you pop up the network view (Ctrl-Shift-E) you can see the associated json call for the year report and its return which today contains 57 records for 2020 and 366 records for `previous year` (including 1-1-2020). Webserver does not consolidate the data to monthly totals, this is all done in the front-end javascript.

**Comment by gizmocuz** [comment]
Still everything is based on the counter values

**Comment by gordonb3** [comment]
Obviously, but that is not the point

**Comment by gizmocuz** [comment]
Yes it is, that is my point. So when the actual counter values are used, there is no rounding.
"IF" a few minutes of the past day where missed, they are taken into account the next day as the counter keeps on ticking.
It is the same when a third party (where you can have a paid subscription for) queries your meter once a day at midnight, this will never be at 00:00:00, they got 1000+ boxes to query.
You might miss a few Watts, but it will be corrected next day, where you might 'again' miss a few watts, but the next day it is taken into account again etc.. 
If we based the daily usage pure on 'last counter value' - 'start of day counter value' and store this in the database only, it could indeed be wrong, be we store the counter values

**Comment by gordonb3** [comment]
No, it isn't, because that's not how the data is processed. It is in fact not even how the data is mined in the first place, because that is a two step operation which requires that they are executed in the right order. This prerequisite becomes uncertain when Domoticz runs on a multicore machine and thus the daily statistics are based on the 23:55 values rather than the 00:00 values.

Webserver corrects the counter values that are stored in the database by subtracting the corresponding usage that is stored in the very same record, thus resetting the time of registration back to the previous 00:00. This is valid because the counter value and the usage reference the same time. The usage however is passed on as is and the javascript in the front-end also does not recalculate these values, which you can clearly see in the month report. In the year report the monthly totals for usage and costs are determined by the sum of these values in the consecutive day records. i.e. there is no reference to counter values between multiple records, nor is there a cost calculation being done on the grand total.

Thus for a 30 day month, the displayed counter values are correct but there is 2.5 hours of usage missing. In my situation this is likely less than a kWh and that does not really bother me, but it is a returning ground for complaints by several users.

**Comment by gizmocuz** [comment]
But the counter values are used in the backend (C++), and the counter values are stored in the database since 9 october 2014
Again, with the counter values we should get a relative good monthly total
That maybe currently the javascript performs the count, and we could improve it by using the counter values, is another thing.

**Comment by gordonb3** [comment]
I understand that, but please have a look here:
https://github.com/domoticz/domoticz/blob/945568532943b639698b6fabdb65c304bbe6213a/main/WebServer.cpp#L15899-L15907

It's not just the counter values that are stored, but also a calculated usage which is in fact required to correct the counter value as this references an incorrect time. You can verify this yourself by querying `multimeter_calendar` for a specific date and compare the value for `Counter1` with the value displayed in the report. Personally I doubt that the value in the database is ever used as is, because of this incorrect time, but I may be wrong.

Thus if our record for 1-1-2020 says `Value1=10258` and `Counter1=7607999` this will cause webserver to output the following json (not listing the other values):
```
c1 | 7597.741
d | 2020-01-01
v | 10.258
```
And that's a problem because `v != return[1].c1 - return[0].c1` (except on single core machines for dates after the commit of PR #3905)

**Comment by FireWizard52** [comment]
@gizmocuz wrote:

> @FireWizard52 , that is strange, as in the report (for the month) the total month usage could/should be calculated by the counter values and it should be correct, it can not be a few Kwh per month/year
For example month usage = first value of next month - first value of month
There is no need to summarize all day values to produce a month total

As an example I will take my consumption for February  2020.
First value of T1 in March 8141.783; First value of T1 in February 7859.334

If I follow this calculation month usage = first value of next month - first value of month

Conclusion Month Usage of February: 282.449
However at the top of February, it is indicated: Total Usage: T1: 280.641 kWh.
And this last value is exactly the sum of the Usage for all (in this case) 29 days.
No need to summarize?

Which value is correct?

Last week I have experienced from Sunday to Thursday correct updates (running on 4 cores) and since Thursday it does not match.anymore. Try again on 1 core.

**Comment by gizmocuz** [comment]
As mentioned, when using the counter values, then this is the correct one
Could be that the web interface is using the sum of the usages and those are rounded, and you get 200 watt more
So this should be rebuild to use the counter
It should not matter if you have 16 cores, if you use the counter values it would be OK

**Comment by gordonb3** [comment]
@gizmocuz I think you're focussing too much on this specific use case. It's not about the counter values that happen to be available with this particular type of meter and thus allow verification of the calculated usage. It's about the fact that the usage calculation for **all** meter types is wrong due to the fact that the data for the 24 hour period that they are supposed to represent is or may be incomplete when the calculation is done.

**Comment by gizmocuz** [comment]
I don't agree with this...
At midnight we make the daily values, this has nothing to do with 1 core, 10 cores or 64 core machines
It 'might' have to do with 'timing' ... like the system might have a clock drift of some minutes
Anyway, back again to the counter values, they are important !...

01-March-2020 (at 00:00am): 1000000
02-March-2020 (at 00:00am): 2000000
01-April-2020 (at 00:00am): 5000000

Usage 01 March = 2000000 - 1000000
Usage March = 5000000 - 1000000

that we take a value on 02-March-2020 (at ~00:00) of 1999999 or 2000001 ... whatever ... 

Be assured that we first perform the shortlog database logging, and then the long turn logging (day values)

The P1 power meter sends an update every 10 seconds... should by quite accurate

**Comment by gordonb3** [comment]
```
Be assured that we first perform the shortlog database logging, and then the long turn logging (day values)
```

That's the point. That's why it runs consistently correct on single core machines (since PR #3905). It doesn't work like that on multicore machines though, because events can run in parallel and thus the midnight event may access the shortlog before the last log entry (the 00:00 value that completes the 24 hour cycle) was added by the 5 minute event.

**Comment by gordonb3** [comment]
Also:

Do note that the `calendar` will not hold the values that you mentioned like that. To start with there is no record for today (March 2). The record for March 1, using your numbers, will show a `[Counter1]` value of 2000000 and a value named `[Value1]` of 1000000 which is the usage.

That is: if everything happens in the right order. In practice you will see records holding values `Counter1 = 1950000` and `Value1 = 950000`. Do note that this still leads to a display of 1000000 for `Counter1` on March 1, however when you add the usage this will not result in the starting value for March 2, which is 2000000. Instead it will give you the March 1 23:55 value for Counter1, because that was the last entry in the shortlog when the calendar record was created.

**Comment by gizmocuz** [comment]
@gordonb3 , no it runs like it is designed, there is one worker thread in mainworker (do_work) that handles both cases

**Comment by gizmocuz** [comment]
@gordonb3 , sure the record of 2 march can not be made until midnight

**Comment by gordonb3** [comment]
Right. Now I'm intrigued even more. If the code doesn't start multiple threads, which appears to be like you say, then why does `AddCalendarUpdateMultiMeter()` fail to retrieve the midnight entry from the shortlog when Domoticz is running on a multithreaded CPU?

Do note that @FireWizard52 reported that the data was correct when processor affinity was set to force Domoticz to use a single core only on the very same machine.

**Comment by gizmocuz** [comment]
It should not... If there was data in the shortlog, it should have no issue at all.
I already found that PR strange but OK...

**Comment by gordonb3** [comment]
No, the PR is correct. The original code only retrieved the data that matched a specific date in the shortlog and thus only evaluated the values from 00:00 up to 23:55. To make it a full 24 hours the 00:00 entry for the following day has to be included.

**Comment by FireWizard52** [comment]
Hi,
I changed yesterday once more Domoticz to 1 core instead of 4 cores and I have to confirm that the update was correct this morning. I also compared the counter values of 2019 with the values provided by the Energy supplier. They are identical, so the counter values are correct as Gizmocuz said.
But compared to the Energy supplier, the sum of all Usages is not correct and that means also that the values presented at the top left corner of the table are wrong. In 2019, in my case, the difference is 63.5 kWh. That is , what I mean with "This depends on how you calculate per day/month/year."
I estimated 40 kWh, but it is 63.5 kWh

**Comment by gizmocuz** [comment]
So again, the counter values are correct... they are taken over from the meter device
We should use the counter values for all calculations
This has to be changed in the code
In October 2014 I added the counter values to the database table
Of course I would still like to see all reports before 2014 as well...

**Comment by gordonb3** [comment]
@gizmocuz Not exactly.

For starters, not all meters have counters and thus redesigning the front-end to reference the counters only won't work as a generic solution.

Second, the back-end **is** sending values for usage and it's kind of silly to state that whatever front-end should not use those because they cannot be trusted.

Third, the counter values as stored in `meter_calendar` and `multimeter_calendar` are in fact unreliable as well because they can either reference the counter values at `23:55:00` or those from `23:59:59.999`. By subtracting the usage that references that exact same time these values can however give an accurate representation for the counter value at `00:00:00`. The amazing thing here is that the counter values that are exported in bulk are the ones that require this setback in time, whereas the procedures that reference the value as it is stored (and is thus not to be trusted) only use the most recent entry.

And that is why I asked for your input. Because as I see it this whole calendar handling is all wrong.

**Comment by gizmocuz** [comment]
@gordonb3 , good meters HAVE counters, but that is besides the OP topic.
I am still talking about P1 here !!
Now let us end this discussion until me (or someone else) are taking the counters into account
This is a pointless discussion as you already might have guessed as I will keep hammering on using the counter values
You can NEVER get YOUR exact counter values as the clock inside the P1 meter is Off, your Domoticz clock might be off bla bla bla
In the end it does not matter, we are talking about 10 seconds here as the P1 meter send it's energy counters every 10 seconds
So let's stop this discussion for now... PLEASE ... i have other matters to attend to that are more pressing
How hard can it be, at 00:00:00am we make the first value for the new day, this value is used together with the first value from the previous day at 00:00:00am and we have a total.. simple math!
And nothing wrong with it..

**Comment by gordonb3** [comment]
@FireWizard52 
Sorry, I tried. But this is what essentially always happens when gizmocuz gets something in his head. Once you go past what his ego is willing to accept, which was probably some four days ago,  it becomes impossible to make him review that thought.

My suggestion is that you wait maybe two months and then open a new issue on this subject. In the meantime you can use the perl script I provided in the forum to fix the numbers in `multimeter_calendar`. Please do not reference me any more with regard to this subject, I will mute all existing and future threads and not respond.

Best of luck,
Gordon

**Comment by gizmocuz** [comment]
@gordonb3 , thanks, I appreciate your comment very much!
(... talking about ego...and for the record it was already 23 days ago)

**Comment by FireWizard52** [comment]
It is very pity to read that two potentially excellent coders are not able to agree on the fact that something is wrong. It is obvious.
As said, the counter values matches the read-outs by the energy supplier. So they are correct, because at the very end we have to pay for what they measure and not what Domoticz measures or calculates.. However the Usage T1 is not correct and that will have effect on the total column in that month and also in the summarized values at the top left corner. And ithas only effect on T1 and not T2, R1 and R2. As a coder it would be below my level of honour to leave it that way, but I'm not a coder.

**Comment by gizmocuz** [comment]
@FireWizard52,

This is what happens internally from the c++ backend to the web gui and why it looks the counter values of T2,R1 and R2 are the same as in the database, but they are not all the time!

WebServer.cpp, arround line 15894

```
								if (counter_1 != 0)
								{
									sprintf(szTmp, "%.3f", (counter_1 - fUsage_1) / divider);
								}
								else
								{
									strcpy(szTmp, "0");
								}

```

counter_1 = 18578582 (In database)
fUsage_1 = 12766 (In database)

result in report = 18578582-12766= **18565.816**

![image](https://user-images.githubusercontent.com/2265299/75789180-96224900-5d69-11ea-8989-3bdc5f86c4c0.png)

The same is done for T2, but notice here that fUsage_2 (usage T2) is 0, so in this case the counter value from the database is also seen in the report

(And the same for R1 / R2)

So that is why you see the shift

**Comment by gizmocuz** [comment]
Do note that when fUsage_2, fDeliv_1 or fDeliv_2 does contain a value, this is subtracted from the counter causing the shift to appear in the report

**Comment by gordonb3** [comment]
Christ...

You just couldn't let it go, could you? You had to mention me so that GitHub would spam me with that stupid pride of yours. And of course you had to show off that you are incapable of accepting any criticism about your work, even though no person ever expressed such a thing within this issue. It was your own paranoid mind that caused you to think that you had to defend that pathetic pride of yours and come up with a bullshit theory that makes no sense at all and an idiot idea that you will never execute yourself nor allow any other person to enter through a PR.

No I don't particularly think of myself as a better person than you, which is why I feel no reserve whatsoever returning your favour by telling you that I actually did figure out what causes the multi-threading problem. Since however your pride forbids you to accept that this problem even exists, you will be extremely happy to learn that I will not submit a PR for it, nor inform you in any other way what the solution is. Not even if the both of you happen to come to your senses. That's you and your ego. Enjoy your evening, go play checkers or whatever.

**Comment by gizmocuz** [comment]
*wow* ... Maybe it's better to take a break and take a few days off to relax dude !
If something is not 100% clear for me i will NOT accept a PR. And yes I am very stubborn, and that is my right.
I am always open for discussions on Slack
Have a nice day!
